blocks/monitoring/__init__.py

```diff
@@ -0,0 +1,193 @@
+"""This part of the framework helps you monitor your model during training.
+
+:class:`MonitorChannels` describe quantities that should be monitored. They
+can either be Theano variables (e.g. the objective function, the weight
+norms, etc.) or they can be callable functions (e.g. if you want to use
+NumPy to calculate eigenvalues, or sample text from your language model).
+
+.. todo::
+
+   Monitored Theano variables might depend on inputs beyond those given to
+   the model. There should be a way of providing these to the monitoring
+   channels, collecting them, and compiling them as part of the Theano
+   function that performs monitoring.
+
+"""
+import logging
+from abc import ABCMeta, abstractmethod
+
+from blocks.utils import shared_expression, update_instance
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorChannel(object):
```

Comment:
> I think datasets should definitely have sources, it worked well in GH 
apparently, and it did in Pylearn2 as well. I was just pointing out that 
this hasn't been implemented yet. Working on that now!

Well GH is very limited - it just assumes that the data iterator yields 
dicts and that all theano input variables' names somehow match the keys 
in the dict. There is no mechanism to select what you want form a 
dataset, which is in PL2. On the other hand PL2 assumes that you will 
have many different sources with the same name, but different format, 
thus the batch cannot be a dict but has to be a tuple that you zip with 
the sources to get to the components. I find GH approach more intuitive, 
precisely because there is no source selection magic like in PL2 and 
there is no assumption that stuff with the same name can be different.